frontend/src/metabase/admin/settings/selectors.js

```diff
@@ -26,20 +26,43 @@ import { PLUGIN_ADMIN_SETTINGS_UPDATES } from "metabase/plugins";
 
 // This allows plugins to update the settings sections
 function updateSectionsWithPlugins(sections) {
-  return PLUGIN_ADMIN_SETTINGS_UPDATES.reduce(
-    (sections, update) => update(sections),
-    sections,
-  );
+  if (PLUGIN_ADMIN_SETTINGS_UPDATES.length > 0) {
+    const reduced = PLUGIN_ADMIN_SETTINGS_UPDATES.reduce(
+      (sections, update) => update(sections),
+      sections,
+    );
+
+    // the update functions may change the key ordering inadvertently
+    // see: https://github.com/aearly/icepick/issues/48
+    // therefore, re-sort the reduced object according to the original key order
+    const reSortFn = (a, b) => {
+      if (sections[a[0]] !== undefined && sections[b[0]] !== undefined) {
+        const aMap = sections[a[0]];
+        const bMap = sections[b[0]];
+        if (aMap.order !== undefined && bMap.order !== undefined) {
+          return aMap.order - bMap.order;
+        }
+      }
+      // undefined relative ordering, so just sort these two arbitrarily
+      return 1;
+    };
+
+    return Object.fromEntries(Object.entries(reduced).sort(reSortFn));
```

Comment:
in general, we shouldn't rely on order in objects since things like this can happen really easily, so we should probably make this an array of objects at some point soon. i won't make you do that here, but we can simplify the sort function a little bit. `Object.entries` generates a list of `[key, value]` pairs, so we can make `reSortFn`:

```
([, aVal], [, bVal]) => aVal && bVal && aVal.order - bVal.order
```

I suppose `aVal` and `bVal` have a chance to be `null` or `undefined` but I wouldn't worry too much about safeguards besides that.

Reply:
Done